ksql-engine/src/main/java/io/confluent/ksql/analyzer/Analyzer.java

```diff
@@ -83,471 +85,80 @@
-    node.getHaving().ifPresent(this::analyzeHaving);
-    node.getLimit().ifPresent(analysis::setLimitClause);
-
-    analyzeExpressions();
+  /**
+   * Analyze the query.
+   *
+   * @param query the query to analyze.
+   * @param sink the sink the query will output to.
+   */
+  void analyze(final Query query, final Optional<Sink> sink) {
+    new Visitor().process(query, null);
 
-    return null;
+    analyzeSink(sink);
   }
 
-  private void analyzeNonStdOutSink(final boolean doCreateInto) {
-    final List<Pair<StructuredDataSource, String>> fromDataSources = analysis.getFromDataSources();
-
-    final StructuredDataSource intoStructuredDataSource = analysis.getInto();
-    final String intoKafkaTopicName = analysis.getIntoKafkaTopicName() == null
-                                      ? topicPrefix + intoStructuredDataSource.getName()
-                                      : analysis.getIntoKafkaTopicName();
-
-    final KsqlTopic newIntoKsqlTopic;
-    if (doCreateInto) {
-      KsqlTopicSerDe intoTopicSerde = fromDataSources.get(0).getLeft().getKsqlTopic()
-          .getKsqlTopicSerDe();
-      if (analysis.getIntoFormat() != null) {
-        switch (analysis.getIntoFormat().toUpperCase()) {
-          case DataSource.AVRO_SERDE_NAME:
-            final String schemaFullName =
-                StringUtil.cleanQuotes(
-                 analysis.getIntoProperties().get(
-                   DdlConfig.VALUE_AVRO_SCHEMA_FULL_NAME).toString());
-            intoTopicSerde = new KsqlAvroTopicSerDe(schemaFullName);
-            break;
-          case DataSource.JSON_SERDE_NAME:
-            intoTopicSerde = new KsqlJsonTopicSerDe();
-            break;
-          case DataSource.DELIMITED_SERDE_NAME:
-            intoTopicSerde = new KsqlDelimitedTopicSerDe();
-            break;
-          default:
-            throw new KsqlException(
-              String.format("Unsupported format: %s", analysis.getIntoFormat()));
-        }
-      } else {
-        if (intoTopicSerde instanceof KsqlAvroTopicSerDe) {
-          final String schemaFullName =
-              StringUtil.cleanQuotes(
-                analysis.getIntoProperties().get(
-                  DdlConfig.VALUE_AVRO_SCHEMA_FULL_NAME).toString());
-          intoTopicSerde = new KsqlAvroTopicSerDe(schemaFullName);
-        }
-      }
-
-      newIntoKsqlTopic = new KsqlTopic(
-          intoStructuredDataSource.getName(),
-          intoKafkaTopicName,
-          intoTopicSerde,
-          true
-      );
-    } else {
-      newIntoKsqlTopic = metaStore.getTopic(intoStructuredDataSource.getName());
-      if (newIntoKsqlTopic == null) {
-        throw new KsqlException(
-            "Sink topic " + intoKafkaTopicName + " does not exist in the metastore.");
-      }
+  private void analyzeSink(final Optional<Sink> sink) {
+    if (sink.isPresent()) {
+      final StructuredDataSource into = analyzeNonStdOutTable(sink.get());
+      analysis.setInto(into, sink.get().shouldCreateSink());
```

Comment:
Can we move this into analyzeNotStdOutSink(), and pass `into` as a parameter, rather than using the analysis to pass it around

Reply:
done. (Though when you start picking at that thread there's a long way to go...